infra/src/main/kotlin/fi/fta/geoviite/api/frameconverter/v1/FrameConverterServiceV1.kt

```diff
@@ -156,6 +186,94 @@ class FrameConverterServiceV1 @Autowired constructor(
         }
     }
 
+    fun validateTrackAddressToCoordinateRequest(
+        request: TrackAddressToCoordinateRequestV1,
+    ): Pair<ValidTrackAddressToCoordinateRequestV1?, List<GeoJsonFeatureErrorResponseV1>> {
+        val validKilometerRange = 0..9999
+        val validMeterRange = 0..9999
```

Comment:
Nämä on vähän hähmeitä, kilometrirajaksi varmaan tuo 0..9999 toimii, mutta mites metrit? Ratakilometrillä voi kuitenkin olla enemmän kuin 1000 metriä, mutta kuinka paljon enemmän? Mihis tuo raja asetettaisiin?

Reply:
Tarkistin, tällä hetkellä pisin ratakilometri tuotannossa on vajaat 3300 metriä. Luulisi että ne eivät tuosta enää kauheasti pitene ainakaan tarkoituksella, joten esim. joku 3500-5000 voisi olla ihan ok saniteettiraja mielestäni

Reply:
Tai no oikeastaan kun tätä mietin, niin tässähän on useampi eri erikoistilanne jotka pitäisi ottaa jotenkin huomioon:
1. Meillähän ei (muistaakseni) mikään taida estää todella pitkien ratakilometrien luomista jos niille vaan löytyy geometria. Pitäisikö tämän kuitenkin toimia myös vaikka 6km (tai 10km tai 100km) pitkille ratakilometreille mikäli sellaisia jostain syystä tehdään paikannuspohjaan?
2. Miten tämän pitäisi toimia tilanteessa, jossa pyydetty metriarvo on isompi kuin ratakilometrin pituus? Pitäisikö metriarvoa käsitellä pelkkänä offsettina annetun ratakilometrin alusta vai arvona jonka kuuluu olla ratakilometrin pituuden sisällä? (Ts. jos 0001-ratakilometri on 1000 metriä pitkä, pitääkö 0001+1500-request käsitellä a) niinkuin se olisi 0002+0500 vai b) palauttamalla virhe?) Jos tätä ei ole speksattu, niin sen voisi tarkistaa vanhasta toteutuksesta. Omaan korvaan heittäminen kuulostaisi järkevämmältä, mutta oikeastaan molemmissa tapauksissa jättäisin tämän range-checkin täältä pois, sillä isoille arvoille löytyy kuitenkin selkeästi mietitty käsittely myöhempää ketjusta

Reply:
Aiempi vkm toiimii samoin kuin tässä pullarissa toteuttu logiikka, eli jos esimerkkiratakilometriltä 0001 haluaa pyytää metriä 1500, niin pyyntö olisi rataosoitteelle 0001+1500, mutta oikeastihan nuo ei ole rataosoitterakenteella edes annettu, vaan kahdella kentällä erikseen (ratakilometri INT ja ratametri INT).

Tuo toimii myös meillä UI:lla näin, ks. esim. ILR 804 ja seuraa millaisia rataosoitearvoja alakulmassa näkyy (esim. 0005+1190).

Reply:
Mikään ei siis sinällään taida estää epäintuitiivista rakennetta, että ratakilometri 0001 on 5000 metriä pitkä ja ratakilometri 0002 on 2024 metriä pitkä, kun tuosta herkästi olettaisi, ettei ratakilometri 0002 voisi alkaa vasta 5000 metriä ensimmäisen ratakilometrin jälkeen.

Reply:
Itse tekisin tämän validoinnin tuottamalla siitä meidän sisäisen tyypin TrackMeter. Jos tuon voi tuottaa, se on validi osoite. Jos tuon luonti päästää jonkun epävalidin läpi, on parempi korjata sen sisäistä validointia. Sinne voi myös halutessaan tehdä jonkun tryCreate-funkkarin jos ei halua virhettä poikkeuksen muodossa ulos, mutta näin saadaan validointi vain yhteen paikkaan.
Tuota tietenkin vaikeuttaa se jos nuo virheen variantit täytyy olla tietyllä tapaa muotoiltu. Sen voisi ehkä hoitaa tyypittämällä TrackMeter-tyypin heittämät errorit sen verran tarkemmin että virheet voi kantaa tarvittavan datan mukanaan.

Reply:
Ja joo, ratakilometrit on ikävästi nimetty kilometreiksi kun ne ei oikeasti ole pituutta vaan osoitteita jotka koittaa vähän joskus imitoida matkaakin. Ne sisäisen validoinnin rajat taisi olla että kilometri ei voi olla yli 10km pitkä koska sillekin on speksissä rajattu vain 4 numeroa.

Reply:
Jeps, TrackMeterin hyödyntäminen vaikuttaisi tässä kyllä ihan mielekkäältä. Se tosin sallii jostain syystä negatiiviset metriarvot, joka ainakin hieman meikäläistä hämää validointia ajatellen:
```        
val asd = TrackMeter("1237+-9999") // toimii normaalisti
```

Nykyinen rata-vkm taas sallii ainoastaan ratametri-arvot välillä 0-2000 eli sieltä ei voi edes kysellä kaikkia nykyisin tuotannossa olevia mahdollisia arvoja.

Esim. Toimiva kysely, jossa ratametri on yli 1000 ja 2000 rajalla:
[https://avoinapi.vaylapilvi.fi/viitekehysmuunnin/muunna?vaylan_luonne=100&ratanumero=352&ratakilometri=430&ratametri=2000&valihaku=false](https://avoinapi.vaylapilvi.fi/viitekehysmuunnin/muunna?vaylan_luonne=100&ratanumero=352&ratakilometri=430&ratametri=2000&valihaku=false)

Esimerkkivirhe kun ratametri on yli 2000:
[https://avoinapi.vaylapilvi.fi/viitekehysmuunnin/muunna?vaylan_luonne=100&ratanumero=352&ratakilometri=430&ratametri=2001&valihaku=false](https://avoinapi.vaylapilvi.fi/viitekehysmuunnin/muunna?vaylan_luonne=100&ratanumero=352&ratakilometri=430&ratametri=2001&valihaku=false)

Meidänhän piti sitä vielä selvittää (ainakin v1+ varten), että halutaanko täsmälleen samanmallisia virheitä, eli en tiedä tarvitaanko tässä vaiheessa toteutettua virhegranulariteettia.

Reply:
Lisäksi nykyinen rata-vkm ei myöskään ilmeisesti tue kyselyjä desimaaliarvoisilla ratametritarkkuuksilla, joita meillä olisi mahdollista tukea (näitä sitten taas tosin palautellaan koordinaatti -> rataosoiterajapinnan yhteydessä ratametri_desimaalit-kenttänä, mutta toiseen suuntaan kysely ei onnistu)

Reply:
Nuo negatiiviset vaikuttaa kyllä bugilta. Periaatteessa joskus on ehkä mahdollista ilmoittaa että "ennen km-tolppaa" ja tuollaista saattaisi olla jossain inframallissa ehkä käytetty, joten lienee joku riski että ne olisi tarkoituksellakin. Tuo pitänee ehkä tarkastaa. Mutta jo tuo +- string-formatointi viittaisi siihen että se on ehkä vain lapsus.

Jos tuohon rata-vkm-käyttöön tarvitsee muuta käyttöä tiukempaa rajaa (kuten että vain tasametrit käy) niin ne voi toki toteuttaa sitten sen tuotetun ratametrin päälle, lisärequirella (esim funkkarilla `TrackMeter.hasIntegerPrecision`)

Reply:
Tällainen on myös sallittu rataosoite: `TrackMeter(-10000000, -9999)`

Vaikuttaa hieman hassulta. Tiketöity tämän muuttelu/tarkistelu https://extranet.vayla.fi/jira/browse/GVT-2757